src/fairseq2/models/llama/integ.py

```diff
@@ -38,3 +53,55 @@ def convert_to_reference_checkpoint(checkpoint: dict[str, Any]) -> dict[str, Any
     }
 
     return convert_model_state_dict(state_dict, key_map)
+
+
+def convert_to_huggingface_config(arch: str, config: LLaMAConfig) -> dict[str, Any]:
+    """Convert Llama's config to a dict mirroring Huggingface's format"""
+
+    def compute_intermediate_size(
+        n: int, ffn_dim_multiplier: float = 1, multiple_of: int = 256
+    ) -> int:
+        """From: https://github.com/huggingface/transformers/blob/82fcac0a7e40dc6cc5e3121d714b9b16775293ad/src/transformers/models/llama/convert_llama_weights_to_hf.py#L171"""
+        return multiple_of * (
+            (int(ffn_dim_multiplier * int(8 * n / 3)) + multiple_of - 1) // multiple_of
+        )
+
+    def is_llama_3_2(arch: str) -> bool:
+        # TODO: this seems too britle
+        return "llama3_2_" in arch
+
+    if config.use_scaled_rope:
+        rope_scaling = {
+            "factor": 32.0 if is_llama_3_2(arch) else 8.0,
+            "low_freq_factor": 1.0,
+            "high_freq_factor": 4.0,
+            "original_max_position_embeddings": 8192,
+            "rope_type": "llama3",
+        }
+    else:
+        # mgleize: not sure of the json.dump behavior if rope_scaling is None
+        rope_scaling = None
+
+    # we only specify the parameters made explicit in the Huggingface converter
+    # https://github.com/huggingface/transformers/blob/93aafdc620d39b9ec714ffecf015a085ea221282/src/transformers/models/llama/convert_llama_weights_to_hf.py#L384
+    return {
+        "architectures": ["Fairseq2LlamaForCausalLM"],
+        "bos_token_id": config.vocab_info.bos_idx,
+        "eos_token_id": config.vocab_info.eos_idx,
```

Comment:
Have you seen any mention of EOT (end-of-turn) token in HG config.json? LLaMA 3 is a bit peculiar and mainly uses EOT for instruction models instead of EOS to separate prompts and responses. Not sure if that is relevant for config.json

Reply:
Thank you for pointing this out to me. After checking there is definitely mention of an instruct-specific eos_token_id behavior in the Huggingface llama conversion script, and it definitely impacts the config.json:

```
if instruct:
    eos_token_id = [128001, 128008, 128009]
else:
    eos_token_id = 128001
```


`instruct` is a command-line flag argument of this script, is there an equivalent in fairseq2? I saw we configure special token ids in VocabularyInfo so the user could set eos_idx to the EOT token?

Reply:
Ok, my bad. Just realized that you get `model_config` based on the model name. Then this should be good to go. For instruct models, we use their model card to make sure that they set the right EOS. You can check it out [here](https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/assets/cards/models/llama.yaml#L76-L80).